src/lib/adapters/WebDav.js

```diff
@@ -0,0 +1,400 @@
+                    method: 'PUT',
+                    headers: {
+                        'Content-Type': 'application/json',
+                        'Authorization': 'Basic ' + btoa(this.server.username + ':' + this.server.password)
+                    },
+                    body: this.bookmarksAsJSON
+                });
+        } catch (e) {
+            console.log ("Error Caught");
+            console.log (e);
+            throw new Error('Network error: Check your network connection and your account details');
+        }
+    }
+
+    async pullFromServer () {
+        let fullUrl = this.server.bookmark_file;
+        fullUrl = this.server.url + fullUrl;
+        console.log ("fullURL :" + fullUrl + ":");
+
+        let myBookmarks = this.getBookmarksAsJSON ();
+        let response;
+
+        try {
+            response = await fetch (fullUrl, {
+                    method: 'GET',
+                    headers: {
+                        'Authorization': 'Basic ' + btoa(this.server.username + ':' + this.server.password)
+                    }
+                });
+        } catch (e) {
+            console.log ("Error Caught");
+            console.log (e);
+            throw new Error('Network error: Check your network connection and your account details');
+        }
+
+        if (response.status === 401) {
+            throw new Error('Couldn\'t authenticate for removing bookmarks from the server.');
+        }
+
+        if (response.status !== 200) {
+            return {
+                'status' : response.status,
+                'db' : new Map ()
+            };
+        }
+
+        let bookmark_array = await response.json();
+        let server_db = new Map ();
+
+        bookmark_array.forEach ( (bm) => {
+            server_db.set(bm.id, {
+                id: bm.id
+                , url: bm.url
+                , title: bm.title
+                , path: bm.path
+            });
+        });
+
+console.log ("response out");
+console.log (server_db);
+
+        return {
+            'status' : response.status,
+            'db' : server_db
+        };
+    }
+
+    async syncStart () {
+        console.log ("syncStart: started");
+        try {
+            let resp = await this.pullFromServer ();
+            this.db = resp.db;
+
+console.log ("resp");
+console.log (resp);
+
+            if (resp.status !== 200)
+            {
+                if (resp.status !== 404)
+                {
+                    throw new Error('Failed to fetch bookmarks :' + resp.status + ":");
+                }
+            }
+        } catch (e) {
+            console.log ("caught error");
+            console.log (e);
+
+            this.db = new Map ();
+        }
+
+        console.log ("syncStart: completed");
+    }
+
+    async pullBookmarks () {
+        console.log('Fetching bookmarks', this.server)
+
+        let server_db;
+        try {
+            let resp = await this.pullFromServer ();
```

Comment:
Why is this called here again? Shouldn't be necessary if syncStart works properly, right?

Reply:
This may need to be looked at.   I was watching the process after many attempts to get webdav working, I may not fully understand this call.   Do you want the current local copy of bookmarks or do you want a pristine copy from the server?   I was not sure.   If the former I can easily just use the local copy from syncStart.

Reply:
If we have a lock file, we can use the one we got at the beginning of the sync. If we get a new one, we lose the changes that happened before pullBookmarks is called, i.e. all SERVERDELETEs and SERVERCREATEs.

Reply:
That makes sense.    Let me look into the lock file mechanisms in webdav.